### ISSUE TITLE
Improve handling of Windows paths

### ISSUE THREAD

**Issue by talex5** [ocaml-multicore/eio#762]
In https://github.com/ocaml-multicore/eio/issues/124, @samwgoldman said:

> For Flow, we have the need to share paths between platforms. These are relative paths, using the `/` path separator. We only support Windows systems that transparently support the `/` path separator, so we can just pass these paths into Windows APIs without transforming them.

However, Eio also needs to support absolute paths on Windows and it seems this doesn't work (see https://discuss.ocaml.org/t/how-to-specify-a-full-windows-path-in-eio/14880). It would be good to investigate why.

@dra27 commented:
> I must confess that introducing yet another notation for a Windows path doesn’t look at all sensible to me! `/C:/path` is simply invalid, and if that’s ever printed (error messages, exceptions, etc.) it kinda just looks like Eio doesn’t know what it’s doing.

As mentioned in the original issue:

> The [Fpath](https://erratique.ch/software/fpath) library encodes lots of knowledge about Windows paths. It probably can't be used directly because its paths behave differently depending on the host platform (preventing e.g. a program running on Linux from creating paths that are to be used on Windows). The path rules should probably be per-filesystem instead.

I think we should give up on a uniform representation of paths and instead move some operations (e.g. `split`) to `Fs.Pi.DIR`, so that each file-system can provide its own implementation.

**Comment by RichardChukwu** [comment]
Thanks for bringing up this issue, @talex5 

I agree that we need to improve how Eio handles Windows paths. Right now, using relative paths with the / separator works for some cases, but we have issues with absolute paths on Windows that need fixing.

**Things to Consider:**
1. Path Format: I understand @dra27's point about not wanting to add another way to represent Windows paths. We want to keep things clear, and a single format can help avoid confusion, especially in error messages.

2. Filesystem-Specific Solutions: It might be a good idea to let each filesystem handle its own path operations. By moving functions like splitting paths to Fs.Pi.DIR, we can keep the API clean and make it easier to support different systems.

3. Using the Fpath Library: We should look into the Fpath library, but we need to make sure it fits our needs and doesn’t create more issues with path handling across different platforms.

**Next Steps:**
Investigation: I’ll start looking into how paths are currently managed in Eio and identify what’s causing problems on Windows.

Community Input: I’d love to hear any ideas from others on how we can solve this.

I’m looking forward to working together to improve Windows path support in Eio!

**Comment by samwgoldman** [comment]
Apologies for the misleading feedback I shared. Absolute paths never came up in Flow (in practice, all paths are relative to the process's working directory) and I wasn't aware that Windows APIs don't support `/` for absolute paths.

**Comment by talex5** [comment]
@samwgoldman not misleading at all; it was very useful to know we could ship 1.0 like that. I just tagged you in case you objected to a change of direction, but I think it won't affect you either way.

**Comment by dra27** [comment]
Various thoughts, partly following up a chat with @patricoferris last week: it sounds as though it's worth separating what EIO needs _internally_ from paths vs what the EIO user sees/provides. From that:
- There is no "root" directory, so in terms of what's displayed / expected from a caller, one can never be provided (`C:\Windows`, `\Windows`, `\\LOCALHOST\c$\Windows` are all flavours of absolute paths on Windows and, on this machine, with a cwd on Drive C:, they all refer to the _same_ physical directory). I think rather than trying to identify a concept of a root directory, it's better instead to let go of the notion entirely i.e. the portable design is to assume that there _is_ no root directory. _Internally_, NT Object Manager Paths could be used to provide a mapping where there is a root directory (allowing, if needed, there always to be a relative path from one directory to another).
- Modelling the separator _internally_ as being `/` everywhere may sound elegant but, especially as EIO is internally using NT APIs using object manager paths (where `/` is **not** the directory separator) it would possibly be better to use a stronger type than relying on a character (e.g. at its extreme, `string list`, but also using abstract/private strings to track the various conversions needed).
- Just for reference, there's a similar issue manipulating directory lists in environment variables which affects opam 2.2 a lot. The best solution I came up with there is tracking lists where both the normalised and raw forms of the directory entries are kept together. In opam, this is done to keep separators consistent - the nightmare in environment variables comes from `:` being an entry separator on Unix and so prohibited in directory names but being the volume separator on Windows, so the representation allows `C:\nasty path "with redundant quoting" with ";" and spaces\bin;C:\Windows\system32` to be internally treated as two entries `C:\nasty path with redundant quoting with ; and spaces\bin` (that's the _actual_ directory name) and `C:\Windows\system32` but also to be re-displayed to the user with the same quoting as was provided (i.e. keeping the double-quotes where they were). A similar idea could be applied to the directory separators here - i.e. if the user provides `C:\Users/DRA\Documents/Wavy\Separators/Are\In/Fashion` then internally having either `C:/Users/DRA/Documents/Wavy/Separators/Are/In/Fashion` or `\??\C:\Users\DRA\Documents\Wavy\Separators\Are\In\Fashion` is fine, but if the directory is for whatever reason displayed or passed back to the user then only the original form should be seen.

**Comment by talex5** [comment]
> I think rather than trying to identify a concept of a root directory, it's better instead to let go of the notion entirely

Indeed, and Eio doesn't actually require this anyway. The directory `fs` refers to the *current* directory. It's just that it allows access to the rest of the file-system from there (so e.g. `fs / ".."` can be used to refer to the parent directory). Possibly it should have been called `cwd_unconfined` or something.

> the representation allows `C:\nasty path "with redundant quoting" with ";" and spaces\bin;C:\Windows\system32` to be internally treated as two entries `C:\nasty path with redundant quoting with ; and spaces\bin` (that's the actual directory name)

I don't quite understand this. Does this mean that Windows doesn't allow quotes in names, but will ignore them if given?

**Comment by talex5** [comment]
Windows paths should work correctly since #916 was merged.
